### PR TITLE
Refined Colour Sensor Effectivity

### PR DESCRIPTION
**Pull request by glowkate**
Added modifiers to alter the RGB values read in by the colour sensor to allow calibration. Fixed colour ranges on hue to spinner colour. Added to the control panel spinner command so that it updates the dashboard and gives the driver the colour that the colour sensor is currently picking up along with the target colour as dictated at the start of the match.

### CODE REVIEW

**Review by LocPham2003 — APPROVED** [review]

**Review by EvanBialo — APPROVED** [review]

File: src/main/java/ca/warp7/frc2020/commands/ControlPanelCommand.java [hunk 1]
```diff
@@ -9,51 +9,79 @@
...
+            SmartDashboard.putString("targetColour", "No Data");
             //Code for no data received yet
         }
+        
+        SpinnerColour colourSensorReading = spinner.getCurrentSpinnerColor();
+        
+        //If the colour sensed is difrent from the current colour, wait 5 cycles. If it stayed the same
+        //change the current colour to the senced colour.
+        if(currentColour != colourSensorReading){
+            if(colourSensorReading == nextColour && colourCount <= 5){
```
Comment (glowmouse) [inline src/main/java/ca/warp7/frc2020/commands/ControlPanelCommand.java:73]: FYI - I think this needed to be >= 5 for the logic to work as advertised.

Right now I think it does have a positive effect.  If you read red, red, blue, red, red, it'll smooth out the blue.  We won't set it because nextColor = red when we see blue.

File: src/main/java/ca/warp7/frc2020/commands/ControlPanelCommand.java [hunk 2]
```diff
@@ -9,51 +9,79 @@
...
+                currentColour = nextColour;
+            }
+            else if(colourSensorReading == nextColour){
+                colourCount++;
+            }
+            else{
+                colourCount = 0;
+                nextColour = colourSensorReading;
+            }
+        }
```
Comment (glowmouse) [inline src/main/java/ca/warp7/frc2020/commands/ControlPanelCommand.java:83]: When I look at this I think, "this is essentially a state machine"   "a state machine is the only good solution I see"  and "I personally find state machines hard to get right"

Some options for getting this kind of code right:

Option 1: Double check the code by making up some test cases and manually running through them.  If you feed this "red red blue blue blue red blue blue" how do the variables change and does it do what you want?  What about "red red blue blue blue yellow yellow".

Option 2:  

Consider using a more formal state machine abstraction.  The coin operated turn-style from https://en.wikipedia.org/wiki/Finite-state_machine is a good example.  The two states here could be:

"the last sensed colour was the same as the currentColor"
"the last sensed colour was different than the currentColor and we're counting" 

Probably more code, but going through the transitions helps get the corner cases right.

Option 3:

You can unit test stuff like this, but it's infrastructure work.  i.e., you need to create mocks for the inputs and outputs - in this case the spinner and SmartDashboard.  IMHO, not the right way to go a few weeks before competition.

File: src/main/java/ca/warp7/frc2020/subsystems/ControlPanelSpinner.java [hunk 3]
```diff
@@ -26,46 +31,62 @@
 
     private static ControlPanelSpinner instance;
 
+    //These are calibration vars, if you're getting too high or
+    //too low a RGB value then what it should be outputting you
+    //can use these to make adustments to the input.
+    private final int redMod = 0;
+    private final int greenMod = -61;
```
Comment (glowmouse) [inline src/main/java/ca/warp7/frc2020/subsystems/ControlPanelSpinner.java:38]: More documentation links, I think

The light sensor:  http://www.revrobotics.com/rev-31-1557/
And it has a datasheet  http://www.revrobotics.com/content/docs/REV-31-1557-DS.pdf
And the datasheet has a datasheet  https://docs.broadcom.com/docs/APDS-9151-DS

Nothing jumps out about why you needed to adjust green so much.  Weird.

File: src/main/java/ca/warp7/frc2020/subsystems/ControlPanelSpinner.java [hunk 4]
```diff
@@ -26,46 +31,62 @@
...
         return (convertHueToColour(HSB[0]));
     }
-
+    
     /*
     Given a HUE from 0-1, return one of the SpinnerColours based on
     a range of values.
     */
-    public SpinnerColour convertHueToColour(float hue) {
+    public SpinnerColour convertHueToColour(float hue){
```
Comment (glowmouse) [inline src/main/java/ca/warp7/frc2020/subsystems/ControlPanelSpinner.java:70]: Documentation suggestion:  https://en.wikipedia.org/wiki/HSL_and_HSV
I think this is what we're measuring with hue,  https://en.wikipedia.org/wiki/File:HSL-HSV_hue_and_chroma.svg

Mapping to degrees so I'm in sync with the pictures,

red range 342 to 360 degrees 
yellow range 0 to 72
green range 144 to 216
blue range 230 to 252.  

I'd expect red to be around 0 degrees, yellow 60 degrees, green 120, and blue 240.  Everything is ballpark; the only thing that stands out is the yellow being from 0 to 72.  0 should be close to red.

**Review by glowmouse — COMMENTED** [review]
Possibly one follow-up action in ControlPanelCommand.java,  colorCount <= 5.
